Crash on opening jumplist picker

### Summary

Helix randomly crashed after I opened the jumplist picker with Ctrl+J.

### Reproduction Steps

<!-- Ideally provide a key sequence and/or asciinema.org recording. --> 

I tried this:

1. `hx`
2. Ctrl+J
3. Crash


### Helix log

<details><summary>~/.cache/helix/helix.log</summary>

```
2022-09-09T15:28:58.160 helix_lsp::transport [ERROR] <- ServerError(-32900): rustfmt exited with:
                           Status: exit status: 101
                           stdout: 
                           stderr: error[E0765]: unterminated double quote string
   --> <stdin>:450:33
    |
450 |       let expected = String::from("dict\nx\nint\");
    |  _________________________________^
451 | | }
    | |__^


2022-09-09T15:28:58.160 helix_view::document [WARN] LSP formatting failed: protocol error: ServerError(-32900): rustfmt exited with:
                           Status: exit status: 101
                           stdout: 
                           stderr: error[E0765]: unterminated double quote string
   --> <stdin>:450:33
    |
450 |       let expected = String::from("dict\nx\nint\");
    |  _________________________________^
451 | | }
    | |__^


2022-09-09T15:30:01.999 helix_term::application [WARN] lsp position out of bounds - Diagnostic { range: Range { start: Position { line: 449, character: 8 }, end: Position { line: 449, character: 16 } }, severity: Some(Warning), code: Some(String("unused_variables")), code_description: None, source: Some("rustc"), message: "unused variable: `expected`\n`#[warn(unused_variables)]` on by default", related_information: Some([DiagnosticRelatedInformation { location: Location { uri: Url { scheme: "file", cannot_be_a_base: false, username: "", password: None, host: None, port: None, path: "/home/jummit/Programming/rust/src/serializer.rs", query: None, fragment: None }, range: Range { start: Position { line: 449, character: 8 }, end: Position { line: 449, character: 16 } } }, message: "if this is intentional, prefix it with an underscore: `_expected`" }]), tags: Some([Unnecessary]), data: None }
2022-09-09T15:30:01.999 helix_term::application [WARN] lsp position out of bounds - Diagnostic { range: Range { start: Position { line: 449, character: 8 }, end: Position { line: 449, character: 16 } }, severity: Some(Hint), code: Some(String("unused_variables")), code_description: None, source: Some("rustc"), message: "if this is intentional, prefix it with an underscore: `_expected`", related_information: Some([DiagnosticRelatedInformation { location: Location { uri: Url { scheme: "file", cannot_be_a_base: false, username: "", password: None, host: None, port: None, path: "/home/jummit/Programming/rust/src/serializer.rs", query: None, fragment: None }, range: Range { start: Position { line: 449, character: 8 }, end: Position { line: 449, character: 16 } } }, message: "original diagnostic" }]), tags: None, data: None }
```

Stack trace:
```
[1]  + 70856 continued  hx
thread 'main' panicked at 'Attempt to slice past end of RopeSlice: slice end 14045, RopeSlice length 11978', /home/jummit/.local/share/cargo/registry/src/github.com-1ecc6299db9ec823/ropey-1.5.0/src/slice.rs:645:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:143:14
   2: ropey::slice::RopeSlice::slice
   3: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::fold
   4: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &mut F>::call_once
   5: <core::iter::adapters::flatten::FlatMap<I,U,F> as core::iter::traits::iterator::Iterator>::next
   6: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
   7: helix_term::commands::jumplist_picker
   8: helix_term::ui::editor::EditorView::handle_keymap_event::{{closure}}
   9: helix_term::ui::editor::EditorView::handle_keymap_event
  10: <helix_term::ui::editor::EditorView as helix_term::compositor::Component>::handle_event
  11: helix_term::compositor::Compositor::handle_event
  12: helix_term::application::Application::handle_terminal_events
  13: helix_term::application::Application::event_loop_until_idle::{{closure}}
  14: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
  15: std::thread::local::LocalKey<T>::with
  16: tokio::park::thread::CachedParkThread::block_on
  17: tokio::runtime::scheduler::multi_thread::MultiThread::block_on
  18: tokio::runtime::Runtime::block_on
  19: hx::main
```

</details>


### Platform

Linux

### Terminal Emulator

qterminal

### Helix Version

helix 22.08.1 (44b44794)

--- a/helix-term/src/commands.rs
+++ b/helix-term/src/commands.rs
@@ -860,6 +860,7 @@ fn align_selections(cx: &mut Context) {
 
     let transaction = Transaction::change(doc.text(), changes.into_iter());
     doc.apply(&transaction, view.id);
+    view.apply(&transaction, doc);
 }
 
 fn goto_window(cx: &mut Context, align: Align) {
@@ -1290,6 +1291,7 @@ fn replace(cx: &mut Context) {
             });
 
             doc.apply(&transaction, view.id);
+            view.apply(&transaction, doc);
         }
     })
 }
@@ -1307,6 +1309,7 @@ where
     });
 
     doc.apply(&transaction, view.id);
+    view.apply(&transaction, doc);
 }
 
 fn switch_case(cx: &mut Context) {
@@ -2113,6 +2116,7 @@ fn delete_selection_impl(cx: &mut Context, op: Operation) {
         (range.from(), range.to(), None)
     });
     doc.apply(&transaction, view.id);
+    view.apply(&transaction, doc);
 
     match op {
         Operation::Delete => {
@@ -2126,7 +2130,7 @@ fn delete_selection_impl(cx: &mut Context, op: Operation) {
 }
 
 #[inline]
-fn delete_selection_insert_mode(doc: &mut Document, view: &View, selection: &Selection) {
+fn delete_selection_insert_mode(doc: &mut Document, view: &mut View, selection: &Selection) {
     let view_id = view.id;
 
     // then delete
@@ -2134,6 +2138,7 @@ fn delete_selection_insert_mode(doc: &mut Document, view: &View, selection: &Sel
         (range.from(), range.to(), None)
     });
     doc.apply(&transaction, view_id);
+    view.apply(&transaction, doc);
 }
 
 fn delete_selection(cx: &mut Context) {
@@ -2230,6 +2235,7 @@ fn append_mode(cx: &mut Context) {
             [(end, end, Some(doc.line_ending.as_str().into()))].into_iter(),
         );
         doc.apply(&transaction, view.id);
+        view.apply(&transaction, doc);
     }
 
     let selection = doc.selection(view.id).clone().transform(|range| {
@@ -2530,6 +2536,7 @@ async fn make_format_callback(
         let view = view_mut!(editor);
         if doc.version() == doc_version {
             doc.apply(&format, view.id);
+            view.apply(&format, doc);
             doc.append_changes_to_history(view.id);
             doc.detect_indent_and_line_ending();
             view.ensure_cursor_in_view(doc, scrolloff);
@@ -2617,6 +2624,7 @@ fn open(cx: &mut Context, open: Open) {
     transaction = transaction.with_selection(Selection::new(ranges, selection.primary_index()));
 
     doc.apply(&transaction, view.id);
+    view.apply(&transaction, doc);
 }
 
 // o inserts a new line after each line with a selection
@@ -2637,7 +2645,7 @@ fn normal_mode(cx: &mut Context) {
     cx.editor.mode = Mode::Normal;
     let (view, doc) = current!(cx.editor);
 
-    try_restore_indent(doc, view.id);
+    try_restore_indent(doc, view);
 
     // if leaving append mode, move cursor back by 1
     if doc.restore_cursor {
@@ -2654,7 +2662,7 @@ fn normal_mode(cx: &mut Context) {
     }
 }
 
-fn try_restore_indent(doc: &mut Document, view_id: ViewId) {
+fn try_restore_indent(doc: &mut Document, view: &mut View) {
     use helix_core::chars::char_is_whitespace;
     use helix_core::Operation;
 
@@ -2673,18 +2681,19 @@ fn try_restore_indent(doc: &mut Document, view_id: ViewId) {
 
     let doc_changes = doc.changes().changes();
     let text = doc.text().slice(..);
-    let range = doc.selection(view_id).primary();
+    let range = doc.selection(view.id).primary();
     let pos = range.cursor(text);
     let line_end_pos = line_end_char_index(&text, range.cursor_line(text));
 
     if inserted_a_new_blank_line(doc_changes, pos, line_end_pos) {
         // Removes tailing whitespaces.
         let transaction =
-            Transaction::change_by_selection(doc.text(), doc.selection(view_id), |range| {
+            Transaction::change_by_selection(doc.text(), doc.selection(view.id), |range| {
                 let line_start_pos = text.line_to_char(range.cursor_line(text));
                 (line_start_pos, pos, None)
             });
-        doc.apply(&transaction, view_id);
+        doc.apply(&transaction, view.id);
+        view.apply(&transaction, doc);
     }
 }
 
@@ -2999,6 +3008,7 @@ pub mod insert {
         let (view, doc) = current!(cx.editor);
         if let Some(t) = transaction {
             doc.apply(&t, view.id);
+            view.apply(&t, doc);
         }
 
         // TODO: need a post insert hook too for certain triggers (autocomplete, signature help, etc)
@@ -3021,6 +3031,7 @@ pub mod insert {
             indent,
         );
         doc.apply(&transaction, view.id);
+        view.apply(&transaction, doc);
     }
 
     pub fn insert_newline(cx: &mut Context) {
@@ -3108,6 +3119,7 @@ pub mod insert {
 
         let (view, doc) = current!(cx.editor);
         doc.apply(&transaction, view.id);
+        view.apply(&transaction, doc);
     }
 
     pub fn delete_char_backward(cx: &mut Context) {
@@ -3202,6 +3214,7 @@ pub mod insert {
             });
         let (view, doc) = current!(cx.editor);
         doc.apply(&transaction, view.id);
+        view.apply(&transaction, doc);
 
         lsp::signature_help_impl(cx, SignatureHelpInvoked::Automatic);
     }
@@ -3220,6 +3233,7 @@ pub mod insert {
                 )
             });
         doc.apply(&transaction, view.id);
+        view.apply(&transaction, doc);
 
         lsp::signature_help_impl(cx, SignatureHelpInvoked::Automatic);
     }
@@ -3413,7 +3427,7 @@ enum Paste {
     Cursor,
 }
 
-fn paste_impl(values: &[String], doc: &mut Document, view: &View, action: Paste, count: usize) {
+fn paste_impl(values: &[String], doc: &mut Document, view: &mut View, action: Paste, count: usize) {
     let repeat = std::iter::repeat(
         values
             .last()
@@ -3457,6 +3471,7 @@ fn paste_impl(values: &[String], doc: &mut Document, view: &View, action: Paste,
         (pos, pos, values.next())
     });
     doc.apply(&transaction, view.id);
+    view.apply(&transaction, doc);
 }
 
 pub(crate) fn paste_bracketed_value(cx: &mut Context, contents: String) {
@@ -3549,6 +3564,7 @@ fn replace_with_yanked(cx: &mut Context) {
             });
 
             doc.apply(&transaction, view.id);
+            view.apply(&transaction, doc);
         }
     }
 }
@@ -3572,6 +3588,7 @@ fn replace_selections_with_clipboard_impl(
             });
 
             doc.apply(&transaction, view.id);
+            view.apply(&transaction, doc);
             doc.append_changes_to_history(view.id);
             Ok(())
         }
@@ -3642,6 +3659,7 @@ fn indent(cx: &mut Context) {
         }),
     );
     doc.apply(&transaction, view.id);
+    view.apply(&transaction, doc);
 }
 
 fn unindent(cx: &mut Context) {
@@ -3681,6 +3699,7 @@ fn unindent(cx: &mut Context) {
     let transaction = Transaction::change(doc.text(), changes.into_iter());
 
     doc.apply(&transaction, view.id);
+    view.apply(&transaction, doc);
 }
 
 fn format_selections(cx: &mut Context) {
@@ -3728,6 +3747,7 @@ fn format_selections(cx: &mut Context) {
         // );
 
         // doc.apply(&transaction, view.id);
+        // view.apply(&transaction, doc);
     }
 }
 
@@ -3783,6 +3803,7 @@ fn join_selections_inner(cx: &mut Context, select_space: bool) {
     };
 
     doc.apply(&transaction, view.id);
+    view.apply(&transaction, doc);
 }
 
 fn keep_or_remove_selections_impl(cx: &mut Context, remove: bool) {
@@ -3936,6 +3957,7 @@ fn toggle_comments(cx: &mut Context) {
     let transaction = comment::toggle_line_comments(doc.text(), doc.selection(view.id), token);
 
     doc.apply(&transaction, view.id);
+    view.apply(&transaction, doc);
     exit_select_mode(cx);
 }
 
@@ -3992,6 +4014,7 @@ fn rotate_selection_contents(cx: &mut Context, direction: Direction) {
     );
 
     doc.apply(&transaction, view.id);
+    view.apply(&transaction, doc);
 }
 
 fn rotate_selection_contents_forward(cx: &mut Context) {
@@ -4488,6 +4511,7 @@ fn surround_add(cx: &mut Context) {
 
         let transaction = Transaction::change(doc.text(), changes.into_iter());
         doc.apply(&transaction, view.id);
+        view.apply(&transaction, doc);
     })
 }
 
@@ -4527,6 +4551,7 @@ fn surround_replace(cx: &mut Context) {
                 }),
             );
             doc.apply(&transaction, view.id);
+            view.apply(&transaction, doc);
         });
     })
 }
@@ -4554,6 +4579,7 @@ fn surround_delete(cx: &mut Context) {
         let transaction =
             Transaction::change(doc.text(), change_pos.into_iter().map(|p| (p, p + 1, None)));
         doc.apply(&transaction, view.id);
+        view.apply(&transaction, doc);
     })
 }
 
@@ -4729,6 +4755,7 @@ fn shell(cx: &mut compositor::Context, cmd: &str, behavior: &ShellBehavior) {
     if behavior != &ShellBehavior::Ignore {
         let transaction = Transaction::change(doc.text(), changes.into_iter());
         doc.apply(&transaction, view.id);
+        view.apply(&transaction, doc);
         doc.append_changes_to_history(view.id);
     }
 
@@ -4792,6 +4819,7 @@ fn add_newline_impl(cx: &mut Context, open: Open) {
 
     let transaction = Transaction::change(text, changes);
     doc.apply(&transaction, view.id);
+    view.apply(&transaction, doc);
 }
 
 /// Increment object under cursor by count.
@@ -4885,6 +4913,7 @@ fn increment_impl(cx: &mut Context, amount: i64) {
         let transaction = transaction.with_selection(selection.clone());
 
         doc.apply(&transaction, view.id);
+        view.apply(&transaction, doc);
     }
 }
 

--- a/helix-term/src/commands.rs
+++ b/helix-term/src/commands.rs
@@ -27,6 +27,7 @@ use helix_core::{
     SmallVec, Tendril, Transaction,
 };
 use helix_view::{
+    apply_transaction,
     clipboard::ClipboardType,
     document::{FormatterError, Mode, SCRATCH_BUFFER_NAME},
     editor::{Action, Motion},
@@ -859,8 +860,7 @@ fn align_selections(cx: &mut Context) {
     changes.sort_unstable_by_key(|(from, _, _)| *from);
 
     let transaction = Transaction::change(doc.text(), changes.into_iter());
-    doc.apply(&transaction, view.id);
-    view.apply(&transaction, doc);
+    apply_transaction(&transaction, doc, view);
 }
 
 fn goto_window(cx: &mut Context, align: Align) {
@@ -1290,8 +1290,7 @@ fn replace(cx: &mut Context) {
                 }
             });
 
-            doc.apply(&transaction, view.id);
-            view.apply(&transaction, doc);
+            apply_transaction(&transaction, doc, view);
         }
     })
 }
@@ -1308,8 +1307,7 @@ where
         (range.from(), range.to(), Some(text))
     });
 
-    doc.apply(&transaction, view.id);
-    view.apply(&transaction, doc);
+    apply_transaction(&transaction, doc, view);
 }
 
 fn switch_case(cx: &mut Context) {
@@ -2115,8 +2113,7 @@ fn delete_selection_impl(cx: &mut Context, op: Operation) {
     let transaction = Transaction::change_by_selection(doc.text(), selection, |range| {
         (range.from(), range.to(), None)
     });
-    doc.apply(&transaction, view.id);
-    view.apply(&transaction, doc);
+    apply_transaction(&transaction, doc, view);
 
     match op {
         Operation::Delete => {
@@ -2131,14 +2128,10 @@ fn delete_selection_impl(cx: &mut Context, op: Operation) {
 
 #[inline]
 fn delete_selection_insert_mode(doc: &mut Document, view: &mut View, selection: &Selection) {
-    let view_id = view.id;
-
-    // then delete
     let transaction = Transaction::change_by_selection(doc.text(), selection, |range| {
         (range.from(), range.to(), None)
     });
-    doc.apply(&transaction, view_id);
-    view.apply(&transaction, doc);
+    apply_transaction(&transaction, doc, view);
 }
 
 fn delete_selection(cx: &mut Context) {
@@ -2234,8 +2227,7 @@ fn append_mode(cx: &mut Context) {
             doc.text(),
             [(end, end, Some(doc.line_ending.as_str().into()))].into_iter(),
         );
-        doc.apply(&transaction, view.id);
-        view.apply(&transaction, doc);
+        apply_transaction(&transaction, doc, view);
     }
 
     let selection = doc.selection(view.id).clone().transform(|range| {
@@ -2535,8 +2527,7 @@ async fn make_format_callback(
         let doc = doc_mut!(editor, &doc_id);
         let view = view_mut!(editor);
         if doc.version() == doc_version {
-            doc.apply(&format, view.id);
-            view.apply(&format, doc);
+            apply_transaction(&format, doc, view);
             doc.append_changes_to_history(view.id);
             doc.detect_indent_and_line_ending();
             view.ensure_cursor_in_view(doc, scrolloff);
@@ -2623,8 +2614,7 @@ fn open(cx: &mut Context, open: Open) {
 
     transaction = transaction.with_selection(Selection::new(ranges, selection.primary_index()));
 
-    doc.apply(&transaction, view.id);
-    view.apply(&transaction, doc);
+    apply_transaction(&transaction, doc, view);
 }
 
 // o inserts a new line after each line with a selection
@@ -2692,8 +2682,7 @@ fn try_restore_indent(doc: &mut Document, view: &mut View) {
                 let line_start_pos = text.line_to_char(range.cursor_line(text));
                 (line_start_pos, pos, None)
             });
-        doc.apply(&transaction, view.id);
-        view.apply(&transaction, doc);
+        apply_transaction(&transaction, doc, view);
     }
 }
 
@@ -3007,8 +2996,7 @@ pub mod insert {
 
         let (view, doc) = current!(cx.editor);
         if let Some(t) = transaction {
-            doc.apply(&t, view.id);
-            view.apply(&t, doc);
+            apply_transaction(&t, doc, view);
         }
 
         // TODO: need a post insert hook too for certain triggers (autocomplete, signature help, etc)
@@ -3030,8 +3018,7 @@ pub mod insert {
             &doc.selection(view.id).clone().cursors(doc.text().slice(..)),
             indent,
         );
-        doc.apply(&transaction, view.id);
-        view.apply(&transaction, doc);
+        apply_transaction(&transaction, doc, view);
     }
 
     pub fn insert_newline(cx: &mut Context) {
@@ -3118,8 +3105,7 @@ pub mod insert {
         transaction = transaction.with_selection(Selection::new(ranges, selection.primary_index()));
 
         let (view, doc) = current!(cx.editor);
-        doc.apply(&transaction, view.id);
-        view.apply(&transaction, doc);
+        apply_transaction(&transaction, doc, view);
     }
 
     pub fn delete_char_backward(cx: &mut Context) {
@@ -3213,8 +3199,7 @@ pub mod insert {
                 }
             });
         let (view, doc) = current!(cx.editor);
-        doc.apply(&transaction, view.id);
-        view.apply(&transaction, doc);
+        apply_transaction(&transaction, doc, view);
 
         lsp::signature_help_impl(cx, SignatureHelpInvoked::Automatic);
     }
@@ -3232,8 +3217,7 @@ pub mod insert {
                     None,
                 )
             });
-        doc.apply(&transaction, view.id);
-        view.apply(&transaction, doc);
+        apply_transaction(&transaction, doc, view);
 
         lsp::signature_help_impl(cx, SignatureHelpInvoked::Automatic);
     }
@@ -3470,8 +3454,7 @@ fn paste_impl(values: &[String], doc: &mut Document, view: &mut View, action: Pa
         };
         (pos, pos, values.next())
     });
-    doc.apply(&transaction, view.id);
-    view.apply(&transaction, doc);
+    apply_transaction(&transaction, doc, view);
 }
 
 pub(crate) fn paste_bracketed_value(cx: &mut Context, contents: String) {
@@ -3563,8 +3546,7 @@ fn replace_with_yanked(cx: &mut Context) {
                 }
             });
 
-            doc.apply(&transaction, view.id);
-            view.apply(&transaction, doc);
+            apply_transaction(&transaction, doc, view);
         }
     }
 }
@@ -3587,8 +3569,7 @@ fn replace_selections_with_clipboard_impl(
                 )
             });
 
-            doc.apply(&transaction, view.id);
-            view.apply(&transaction, doc);
+            apply_transaction(&transaction, doc, view);
             doc.append_changes_to_history(view.id);
             Ok(())
         }
@@ -3658,8 +3639,7 @@ fn indent(cx: &mut Context) {
             Some((pos, pos, Some(indent.clone())))
         }),
     );
-    doc.apply(&transaction, view.id);
-    view.apply(&transaction, doc);
+    apply_transaction(&transaction, doc, view);
 }
 
 fn unindent(cx: &mut Context) {
@@ -3698,8 +3678,7 @@ fn unindent(cx: &mut Context) {
 
     let transaction = Transaction::change(doc.text(), changes.into_iter());
 
-    doc.apply(&transaction, view.id);
-    view.apply(&transaction, doc);
+    apply_transaction(&transaction, doc, view);
 }
 
 fn format_selections(cx: &mut Context) {
@@ -3746,8 +3725,7 @@ fn format_selections(cx: &mut Context) {
         //     language_server.offset_encoding(),
         // );
 
-        // doc.apply(&transaction, view.id);
-        // view.apply(&transaction, doc);
+        // apply_transaction(&transaction, doc, view);
     }
 }
 
@@ -3802,8 +3780,7 @@ fn join_selections_inner(cx: &mut Context, select_space: bool) {
         Transaction::change(doc.text(), changes.into_iter())
     };
 
-    doc.apply(&transaction, view.id);
-    view.apply(&transaction, doc);
+    apply_transaction(&transaction, doc, view);
 }
 
 fn keep_or_remove_selections_impl(cx: &mut Context, remove: bool) {
@@ -3956,8 +3933,7 @@ fn toggle_comments(cx: &mut Context) {
         .map(|tc| tc.as_ref());
     let transaction = comment::toggle_line_comments(doc.text(), doc.selection(view.id), token);
 
-    doc.apply(&transaction, view.id);
-    view.apply(&transaction, doc);
+    apply_transaction(&transaction, doc, view);
     exit_select_mode(cx);
 }
 
@@ -4013,8 +3989,7 @@ fn rotate_selection_contents(cx: &mut Context, direction: Direction) {
             .map(|(range, fragment)| (range.from(), range.to(), Some(fragment))),
     );
 
-    doc.apply(&transaction, view.id);
-    view.apply(&transaction, doc);
+    apply_transaction(&transaction, doc, view);
 }
 
 fn rotate_selection_contents_forward(cx: &mut Context) {
@@ -4510,8 +4485,7 @@ fn surround_add(cx: &mut Context) {
         }
 
         let transaction = Transaction::change(doc.text(), changes.into_iter());
-        doc.apply(&transaction, view.id);
-        view.apply(&transaction, doc);
+        apply_transaction(&transaction, doc, view);
     })
 }
 
@@ -4550,8 +4524,7 @@ fn surround_replace(cx: &mut Context) {
                     (pos, pos + 1, Some(t))
                 }),
             );
-            doc.apply(&transaction, view.id);
-            view.apply(&transaction, doc);
+            apply_transaction(&transaction, doc, view);
         });
     })
 }
@@ -4578,8 +4551,7 @@ fn surround_delete(cx: &mut Context) {
 
         let transaction =
             Transaction::change(doc.text(), change_pos.into_iter().map(|p| (p, p + 1, None)));
-        doc.apply(&transaction, view.id);
-        view.apply(&transaction, doc);
+        apply_transaction(&transaction, doc, view);
     })
 }
 
@@ -4754,8 +4726,7 @@ fn shell(cx: &mut compositor::Context, cmd: &str, behavior: &ShellBehavior) {
 
     if behavior != &ShellBehavior::Ignore {
         let transaction = Transaction::change(doc.text(), changes.into_iter());
-        doc.apply(&transaction, view.id);
-        view.apply(&transaction, doc);
+        apply_transaction(&transaction, doc, view);
         doc.append_changes_to_history(view.id);
     }
 
@@ -4818,8 +4789,7 @@ fn add_newline_impl(cx: &mut Context, open: Open) {
     });
 
     let transaction = Transaction::change(text, changes);
-    doc.apply(&transaction, view.id);
-    view.apply(&transaction, doc);
+    apply_transaction(&transaction, doc, view);
 }
 
 /// Increment object under cursor by count.
@@ -4912,8 +4882,7 @@ fn increment_impl(cx: &mut Context, amount: i64) {
         let transaction = Transaction::change(doc.text(), changes);
         let transaction = transaction.with_selection(selection.clone());
 
-        doc.apply(&transaction, view.id);
-        view.apply(&transaction, doc);
+        apply_transaction(&transaction, doc, view);
     }
 }
 

--- a/helix-term/src/commands/lsp.rs
+++ b/helix-term/src/commands/lsp.rs
@@ -596,9 +596,7 @@ pub fn apply_workspace_edit(
             }
         };
 
-        let doc = editor
-            .document_mut(doc_id)
-            .expect("Document for document_changes not found");
+        let doc = doc_mut!(editor, &doc_id);
 
         // Need to determine a view for apply/append_changes_to_history
         let selections = doc.selections();
@@ -620,6 +618,7 @@ pub fn apply_workspace_edit(
             offset_encoding,
         );
         doc.apply(&transaction, view_id);
+        view_mut!(editor, view_id).apply(&transaction, doc);
         doc.append_changes_to_history(view_id);
     };
 

--- a/helix-term/src/commands/lsp.rs
+++ b/helix-term/src/commands/lsp.rs
@@ -9,7 +9,7 @@ use tui::text::{Span, Spans};
 use super::{align_view, push_jump, Align, Context, Editor, Open};
 
 use helix_core::{path, Selection};
-use helix_view::{editor::Action, theme::Style};
+use helix_view::{apply_transaction, editor::Action, theme::Style};
 
 use crate::{
     compositor::{self, Compositor},
@@ -617,8 +617,7 @@ pub fn apply_workspace_edit(
             text_edits,
             offset_encoding,
         );
-        doc.apply(&transaction, view_id);
-        view_mut!(editor, view_id).apply(&transaction, doc);
+        apply_transaction(&transaction, doc, view_mut!(editor, view_id));
         doc.append_changes_to_history(view_id);
     };
 

--- a/helix-term/src/commands/typed.rs
+++ b/helix-term/src/commands/typed.rs
@@ -2,7 +2,10 @@ use std::ops::Deref;
 
 use super::*;
 
-use helix_view::editor::{Action, CloseError, ConfigEvent};
+use helix_view::{
+    apply_transaction,
+    editor::{Action, CloseError, ConfigEvent},
+};
 use ui::completers::{self, Completer};
 
 #[derive(Clone)]
@@ -462,8 +465,7 @@ fn set_line_ending(
             }
         }),
     );
-    doc.apply(&transaction, view.id);
-    view.apply(&transaction, doc);
+    apply_transaction(&transaction, doc, view);
     doc.append_changes_to_history(view.id);
 
     Ok(())
@@ -884,8 +886,7 @@ fn replace_selections_with_clipboard_impl(
                 (range.from(), range.to(), Some(contents.as_str().into()))
             });
 
-            doc.apply(&transaction, view.id);
-            view.apply(&transaction, doc);
+            apply_transaction(&transaction, doc, view);
             doc.append_changes_to_history(view.id);
             Ok(())
         }
@@ -1400,8 +1401,7 @@ fn sort_impl(
             .map(|(s, fragment)| (s.from(), s.to(), Some(fragment))),
     );
 
-    doc.apply(&transaction, view.id);
-    view.apply(&transaction, doc);
+    apply_transaction(&transaction, doc, view);
     doc.append_changes_to_history(view.id);
 
     Ok(())
@@ -1445,8 +1445,7 @@ fn reflow(
         (range.from(), range.to(), Some(reflowed_text))
     });
 
-    doc.apply(&transaction, view.id);
-    view.apply(&transaction, doc);
+    apply_transaction(&transaction, doc, view);
     doc.append_changes_to_history(view.id);
     view.ensure_cursor_in_view(doc, scrolloff);
 

--- a/helix-term/src/commands/typed.rs
+++ b/helix-term/src/commands/typed.rs
@@ -463,6 +463,7 @@ fn set_line_ending(
         }),
     );
     doc.apply(&transaction, view.id);
+    view.apply(&transaction, doc);
     doc.append_changes_to_history(view.id);
 
     Ok(())
@@ -884,6 +885,7 @@ fn replace_selections_with_clipboard_impl(
             });
 
             doc.apply(&transaction, view.id);
+            view.apply(&transaction, doc);
             doc.append_changes_to_history(view.id);
             Ok(())
         }
@@ -1004,7 +1006,7 @@ fn reload(
 
     let scrolloff = cx.editor.config().scrolloff;
     let (view, doc) = current!(cx.editor);
-    doc.reload(view.id).map(|_| {
+    doc.reload(view).map(|_| {
         view.ensure_cursor_in_view(doc, scrolloff);
     })
 }
@@ -1399,6 +1401,7 @@ fn sort_impl(
     );
 
     doc.apply(&transaction, view.id);
+    view.apply(&transaction, doc);
     doc.append_changes_to_history(view.id);
 
     Ok(())
@@ -1443,6 +1446,7 @@ fn reflow(
     });
 
     doc.apply(&transaction, view.id);
+    view.apply(&transaction, doc);
     doc.append_changes_to_history(view.id);
     view.ensure_cursor_in_view(doc, scrolloff);
 

--- a/helix-term/src/ui/completion.rs
+++ b/helix-term/src/ui/completion.rs
@@ -143,11 +143,11 @@ impl Completion {
             let (view, doc) = current!(editor);
 
             // if more text was entered, remove it
-            doc.restore(view.id);
+            doc.restore(view);
 
             match event {
                 PromptEvent::Abort => {
-                    doc.restore(view.id);
+                    doc.restore(view);
                     editor.last_completion = None;
                 }
                 PromptEvent::Update => {
@@ -165,6 +165,7 @@ impl Completion {
                     // initialize a savepoint
                     doc.savepoint();
                     doc.apply(&transaction, view.id);
+                    view.apply(&transaction, doc);
 
                     editor.last_completion = Some(CompleteAction {
                         trigger_offset,
@@ -184,6 +185,7 @@ impl Completion {
                     );
 
                     doc.apply(&transaction, view.id);
+                    view.apply(&transaction, doc);
 
                     editor.last_completion = Some(CompleteAction {
                         trigger_offset,
@@ -214,6 +216,7 @@ impl Completion {
                                 offset_encoding, // TODO: should probably transcode in Client
                             );
                             doc.apply(&transaction, view.id);
+                            view.apply(&transaction, doc);
                         }
                     }
                 }

--- a/helix-term/src/ui/completion.rs
+++ b/helix-term/src/ui/completion.rs
@@ -1,5 +1,5 @@
 use crate::compositor::{Component, Context, Event, EventResult};
-use helix_view::editor::CompleteAction;
+use helix_view::{apply_transaction, editor::CompleteAction};
 use tui::buffer::Buffer as Surface;
 use tui::text::Spans;
 
@@ -164,8 +164,7 @@ impl Completion {
 
                     // initialize a savepoint
                     doc.savepoint();
-                    doc.apply(&transaction, view.id);
-                    view.apply(&transaction, doc);
+                    apply_transaction(&transaction, doc, view);
 
                     editor.last_completion = Some(CompleteAction {
                         trigger_offset,
@@ -184,8 +183,7 @@ impl Completion {
                         trigger_offset,
                     );
 
-                    doc.apply(&transaction, view.id);
-                    view.apply(&transaction, doc);
+                    apply_transaction(&transaction, doc, view);
 
                     editor.last_completion = Some(CompleteAction {
                         trigger_offset,
@@ -215,8 +213,7 @@ impl Completion {
                                 additional_edits.clone(),
                                 offset_encoding, // TODO: should probably transcode in Client
                             );
-                            doc.apply(&transaction, view.id);
-                            view.apply(&transaction, doc);
+                            apply_transaction(&transaction, doc, view);
                         }
                     }
                 }

--- a/helix-term/src/ui/editor.rs
+++ b/helix-term/src/ui/editor.rs
@@ -988,7 +988,7 @@ impl EditorView {
                         InsertEvent::CompletionApply(compl) => {
                             let (view, doc) = current!(cxt.editor);
 
-                            doc.restore(view.id);
+                            doc.restore(view);
 
                             let text = doc.text().slice(..);
                             let cursor = doc.selection(view.id).primary().cursor(text);
@@ -1003,6 +1003,7 @@ impl EditorView {
                                 }),
                             );
                             doc.apply(&tx, view.id);
+                            view.apply(&tx, doc);
                         }
                         InsertEvent::TriggerCompletion => {
                             let (_, doc) = current!(cxt.editor);

--- a/helix-term/src/ui/editor.rs
+++ b/helix-term/src/ui/editor.rs
@@ -16,6 +16,7 @@ use helix_core::{
     visual_coords_at_pos, LineEnding, Position, Range, Selection, Transaction,
 };
 use helix_view::{
+    apply_transaction,
     document::{Mode, SCRATCH_BUFFER_NAME},
     editor::{CompleteAction, CursorShapeConfig},
     graphics::{Color, CursorKind, Modifier, Rect, Style},
@@ -1002,8 +1003,7 @@ impl EditorView {
                                     (shift_position(start), shift_position(end), t)
                                 }),
                             );
-                            doc.apply(&tx, view.id);
-                            view.apply(&tx, doc);
+                            apply_transaction(&tx, doc, view);
                         }
                         InsertEvent::TriggerCompletion => {
                             let (_, doc) = current!(cxt.editor);

--- a/helix-view/src/document.rs
+++ b/helix-view/src/document.rs
@@ -24,7 +24,7 @@ use helix_core::{
     DEFAULT_LINE_ENDING,
 };
 
-use crate::{DocumentId, Editor, View, ViewId};
+use crate::{apply_transaction, DocumentId, Editor, View, ViewId};
 
 /// 8kB of buffer space for encoding and decoding `Rope`s.
 const BUF_SIZE: usize = 8192;
@@ -617,8 +617,7 @@ impl Document {
         // This is not considered a modification of the contents of the file regardless
         // of the encoding.
         let transaction = helix_core::diff::compare_ropes(self.text(), &rope);
-        self.apply(&transaction, view.id);
-        view.apply(&transaction, self);
+        apply_transaction(&transaction, self, view);
         self.append_changes_to_history(view.id);
         self.reset_modified();
 
@@ -811,6 +810,9 @@ impl Document {
     }
 
     /// Apply a [`Transaction`] to the [`Document`] to change its text.
+    /// Instead of calling this function directly, use [crate::apply_transaction]
+    /// to ensure that the transaction is applied to the appropriate [`View`] as
+    /// well.
     pub fn apply(&mut self, transaction: &Transaction, view_id: ViewId) -> bool {
         // store the state just before any changes are made. This allows us to undo to the
         // state just before a transaction was applied.
@@ -865,8 +867,7 @@ impl Document {
 
     pub fn restore(&mut self, view: &mut View) {
         if let Some(revert) = self.savepoint.take() {
-            self.apply(&revert, view.id);
-            view.apply(&revert, self);
+            apply_transaction(&revert, self, view);
         }
     }
 

--- a/helix-view/src/document.rs
+++ b/helix-view/src/document.rs
@@ -24,7 +24,7 @@ use helix_core::{
     DEFAULT_LINE_ENDING,
 };
 
-use crate::{DocumentId, Editor, ViewId};
+use crate::{DocumentId, Editor, View, ViewId};
 
 /// 8kB of buffer space for encoding and decoding `Rope`s.
 const BUF_SIZE: usize = 8192;
@@ -601,7 +601,7 @@ impl Document {
     }
 
     /// Reload the document from its path.
-    pub fn reload(&mut self, view_id: ViewId) -> Result<(), Error> {
+    pub fn reload(&mut self, view: &mut View) -> Result<(), Error> {
         let encoding = &self.encoding;
         let path = self.path().filter(|path| path.exists());
 
@@ -617,8 +617,9 @@ impl Document {
         // This is not considered a modification of the contents of the file regardless
         // of the encoding.
         let transaction = helix_core::diff::compare_ropes(self.text(), &rope);
-        self.apply(&transaction, view_id);
-        self.append_changes_to_history(view_id);
+        self.apply(&transaction, view.id);
+        view.apply(&transaction, self);
+        self.append_changes_to_history(view.id);
         self.reset_modified();
 
         self.detect_indent_and_line_ending();
@@ -862,9 +863,10 @@ impl Document {
         self.savepoint = Some(Transaction::new(self.text()));
     }
 
-    pub fn restore(&mut self, view_id: ViewId) {
+    pub fn restore(&mut self, view: &mut View) {
         if let Some(revert) = self.savepoint.take() {
-            self.apply(&revert, view_id);
+            self.apply(&revert, view.id);
+            view.apply(&revert, self);
         }
     }
 

--- a/helix-view/src/lib.rs
+++ b/helix-view/src/lib.rs
@@ -64,6 +64,19 @@ pub fn align_view(doc: &Document, view: &mut View, align: Align) {
     view.offset.row = line.saturating_sub(relative);
 }
 
+/// Applies a [`helix_core::Transaction`] to the given [`Document`]
+/// and [`View`].
+pub fn apply_transaction(
+    transaction: &helix_core::Transaction,
+    doc: &mut Document,
+    view: &mut View,
+) -> bool {
+    // This is a short function but it's easy to call `Document::apply`
+    // without calling `View::apply` or in the wrong order. The transaction
+    // must be applied to the document before the view.
+    doc.apply(transaction, view.id) && view.apply(transaction, doc)
+}
+
 pub use document::Document;
 pub use editor::Editor;
 pub use theme::Theme;

--- a/helix-view/src/view.rs
+++ b/helix-view/src/view.rs
@@ -3,7 +3,9 @@ use crate::{
     gutter::{self, Gutter},
     Document, DocumentId, ViewId,
 };
-use helix_core::{pos_at_visual_coords, visual_coords_at_pos, Position, RopeSlice, Selection};
+use helix_core::{
+    pos_at_visual_coords, visual_coords_at_pos, Position, RopeSlice, Selection, Transaction,
+};
 
 use std::fmt;
 
@@ -61,6 +63,22 @@ impl JumpList {
 
     pub fn get(&self) -> &[Jump] {
         &self.jumps
+    }
+
+    /// Applies a [`Transaction`] of changes to the jumplist.
+    /// This is necessary to ensure that changes to documents do not leave jump-list
+    /// selections pointing to parts of the text which no longer exist.
+    fn apply(&mut self, transaction: &Transaction, doc: &Document) {
+        let text = doc.text().slice(..);
+
+        for (doc_id, selection) in &mut self.jumps {
+            if doc.id() == *doc_id {
+                *selection = selection
+                    .clone()
+                    .map(transaction.changes())
+                    .ensure_invariants(text);
+            }
+        }
     }
 }
 
@@ -334,6 +352,14 @@ impl View {
     //         (None, None) => return,
     //     }
     // }
+
+    /// Applies a [`Transaction`] to the view.
+    /// Instead of calling this function directly, use [crate::apply_transaction]
+    /// which applies a transaction to the [`Document`] and view together.
+    pub fn apply(&mut self, transaction: &Transaction, doc: &Document) -> bool {
+        self.jumps.apply(transaction, doc);
+        true
+    }
 }
 
 #[cfg(test)]

